src/github.com/getlantern/balancer/dialer.go

```diff
@@ -127,61 +135,40 @@ func (d *dialer) updateEMADialTime(t time.Duration) {
 	// Ref dialer.EMADialTime() for the rationale.
 	// The values is large enough to safely ignore decimals.
 	newEMA := (atomic.LoadInt64(&d.emaDialTime) + t.Nanoseconds()) / 2
-	log.Tracef("Dialer %s EMA(exponential moving average) dial time: %d", d.Label, newEMA)
+	log.Tracef("Dialer %s EMA dial time: %v", d.Label, time.Duration(newEMA))
 	atomic.StoreInt64(&d.emaDialTime, newEMA)
 }
 
 func (d *dialer) markSuccess() {
 	newCS := atomic.AddInt32(&d.consecSuccesses, 1)
 	log.Tracef("Dialer %s consecutive successes: %d -> %d", d.Label, newCS-1, newCS)
-	atomic.StoreInt32(&d.consecFailures, 0)
-	d.muCheckTimer.Lock()
-	d.checkTimer.Reset(maxCheckTimeout)
-	d.muCheckTimer.Unlock()
+	// only when state is changing
+	if newCS <= 2 {
+		atomic.StoreInt32(&d.consecFailures, 0)
+	}
 }
 
 func (d *dialer) markFailure() {
-	atomic.StoreInt32(&d.consecSuccesses, 0)
 	newCF := atomic.AddInt32(&d.consecFailures, 1)
 	log.Tracef("Dialer %s consecutive failures: %d -> %d", d.Label, newCF-1, newCF)
-	nextCheck := time.Duration(newCF*newCF) * 100 * time.Millisecond
-	if nextCheck > maxCheckTimeout {
-		nextCheck = maxCheckTimeout
+	// Don't bother to recheck if dialer is constantly failing.
+	// Balancer will recheck when there's traffic after idle for some time.
+	if newCF < ServerBlacklistingThreshold/2 {
+		atomic.StoreInt32(&d.consecSuccesses, 0)
+		nextCheck := randomize(time.Duration(newCF*newCF) * nextCheckFactor)
```

Comment:
When you run tests, you'll see that `nextCheck` will rapidly increase with `newCF`, even without the dialer actually has a chance to be checked. I think it's fine. More consecutive failures mean that it's more likely to fail again recently, while only one consecutive failure may be temporary network glitch.

If network temporarily goes down for a few seconds and comes up again, there can be a maximum of 5(`ServerBlacklistingThreshold/2`)*5*10 (`nextCheckFactor`) = 250 seconds delay for a dialer to refresh its metrics if other dialers take precedence. It should rarely happen as in that situation, other dialers also fail.